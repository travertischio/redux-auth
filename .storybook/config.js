import React from 'react';
import PropTypes from 'prop-types';
import 'babel-polyfill';
import { fromJS } from 'immutable';
import {
  configure,
  addDecorator,
} from '@kadira/storybook';
import { IntlProvider } from 'react-intl';
import {
  createStore,
  applyMiddleware,
  compose,
} from 'redux';
import { combineReducers } from 'redux-immutable';
import createSagaMiddleware from 'redux-saga';
import {
  Provider,
  connect,
} from 'react-redux';
import { reducer as formReducer } from 'redux-form/immutable';
import { setBaseUrl } from 'api-client';
import { createStructuredSelector } from 'reselect';
import {
  AuthenticationProvider,
  reducer as authReducer,
  sagas as authSagas,
  selectIsAuthenticated,
} from '../src/index';
import signInPageReducer from '../src/containers/SignInPage/reducer';
import signInPageSagas from '../src/containers/SignInPage/sagas';
import signUpPageReducer from '../src/containers/SignUpPage/reducer';
import signUpPageSagas from '../src/containers/SignUpPage/sagas';
import resetPasswordPageReducer from '../src/containers/ResetPasswordPage/reducer';
import resetPasswordPageSagas from '../src/containers/ResetPasswordPage/sagas';
import requestPasswordResetPageReducer from '../src/containers/RequestPasswordResetPage/reducer';
import requestPasswordResetPageSagas from '../src/containers/RequestPasswordResetPage/sagas';
import signOutPageSagas from '../src/containers/SignOutPage/sagas';

setBaseUrl('https://demo-api-dev.arabel.la/api/');

const reducers = {
  form: formReducer,
  auth: authReducer,
  signInPage: signInPageReducer,
  signUpPage: signUpPageReducer,
  resetPasswordPage: resetPasswordPageReducer,
  requestPasswordResetPage: requestPasswordResetPageReducer,
};
const sagaMiddleware = createSagaMiddleware();
const reducer = combineReducers(reducers);
const store = createStore(
  reducer,
  fromJS({}),
  compose(
    applyMiddleware(sagaMiddleware),
    window.devToolsExtension ? window.devToolsExtension() : (f) => f
  )
);

authSagas.map(sagaMiddleware.run);
signInPageSagas.map(sagaMiddleware.run);
signUpPageSagas.map(sagaMiddleware.run);
signOutPageSagas.map(sagaMiddleware.run);
resetPasswordPageSagas.map(sagaMiddleware.run);
requestPasswordResetPageSagas.map(sagaMiddleware.run);

let IsAuthenticatedWrapper = (props) => {
  const {
    isAuthenticated,
    story,
  } = props;

  return (
    <main>
      {isAuthenticated && <h1>You are authenticated!</h1>}
      {!isAuthenticated && <h1>You are not authenticated!</h1>}

      <hr/>

      {story()}
    </main>
  );
};

IsAuthenticatedWrapper.propTypes = {
  isAuthenticated: PropTypes.bool.isRequired,
  story: PropTypes.func.isRequired,
};

const mapStateToProps = createStructuredSelector({
  isAuthenticated: selectIsAuthenticated,
});

IsAuthenticatedWrapper = connect(mapStateToProps)(IsAuthenticatedWrapper);

addDecorator((story) => (
  <IsAuthenticatedWrapper story={story} />
));

addDecorator((story) => (
  <AuthenticationProvider>
    {story()}
  </AuthenticationProvider>
));

addDecorator((story) => (
  <Provider store={store}>
    {story()}
  </Provider>
));

addDecorator((story) => (
  <IntlProvider locale="en">
    {story()}
  </IntlProvider>
));

const req = require.context('../src/', true, /\.stories\.js?$/);

function loadStories() {
  req.keys().forEach((filename) => req(filename));
}

configure(loadStories, module);
