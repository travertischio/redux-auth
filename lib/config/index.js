'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setConfig = undefined;

var _creactUserIsNotAuthenticatedAuthWrapper = require('../hocs/AuthWrappers/creactUserIsNotAuthenticatedAuthWrapper');

var _creactUserIsNotAuthenticatedAuthWrapper2 = _interopRequireDefault(_creactUserIsNotAuthenticatedAuthWrapper);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var config = {
  userIsNotAuthenticatedRedirectPath: '/sign-in',
  userIsNotAdminRedirectPath: '/',
  userHasNoRoleRedirectPath: '/',
  userIsAuthenticatedRedirectPath: '/',
  redirectPathAfterSignIn: '/',
  redirectPathAfterSignUp: '/',
  redirectPathAfterSignOut: '/sign-in',
  adminRole: '20_example_admin',
  camelizeUserDataKeys: false
};

config.signInAuthWrapper = (0, _creactUserIsNotAuthenticatedAuthWrapper2.default)(config.redirectPathAfterSignIn);
config.signUpAuthWrapper = (0, _creactUserIsNotAuthenticatedAuthWrapper2.default)(config.redirectPathAfterSignUp);

var setConfig = exports.setConfig = function setConfig(newConfig) {
  Object.assign(config, newConfig);

  if (newConfig.redirectPathAfterSignIn) {
    updateSignInAuthWrapper();
  }

  if (newConfig.redirectPathAfterSignUp) {
    updateSignUpAuthWrapper();
  }
};

function updateSignInAuthWrapper() {
  config.signInAuthWrapper = (0, _creactUserIsNotAuthenticatedAuthWrapper2.default)(config.redirectPathAfterSignIn);
}

function updateSignUpAuthWrapper() {
  config.signUpAuthWrapper = (0, _creactUserIsNotAuthenticatedAuthWrapper2.default)(config.redirectPathAfterSignUp);
}

updateSignInAuthWrapper();
updateSignUpAuthWrapper();

exports.default = config;