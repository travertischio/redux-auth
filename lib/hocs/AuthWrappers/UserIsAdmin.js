'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _reduxAuthWrapper = require('redux-auth-wrapper');

var _reactRouterRedux = require('react-router-redux');

var _selectors = require('redux-auth/lib/containers/AuthenticationProvider/selectors');

var UserIsAdmin = (0, _reduxAuthWrapper.UserAuthWrapper)({
  authSelector: _selectors.selectUser,
  predicate: isAdmin,
  redirectAction: _reactRouterRedux.routerActions.replace,
  // TODO: move url of default homepage/dashboard to the config file
  failureRedirectPath: '/home',
  allowRedirectBack: false,
  wrapperDisplayName: 'UserIsAdmin'
});
// TODO: move it to redux-auth


function isAdmin(user) {
  return user && user.role === '20_example_admin';
}

exports.default = UserIsAdmin;