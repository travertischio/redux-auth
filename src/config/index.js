import creactUserIsNotAuthenticatedAuthWrapper from '../hocs/AuthWrappers/creactUserIsNotAuthenticatedAuthWrapper';
import UserIsNotAuthenticated from '../hocs/AuthWrappers/UserIsNotAuthenticated';

const config = {
  userIsNotAuthenticatedRedirectPath: '/sign-in',
  userIsNotAdminRedirectPath: '/',
  userHasNoRoleRedirectPath: '/',
  userIsAuthenticatedRedirectPath: '/',
  redirectPathAfterSignIn: '/',
  redirectPathAfterSignUp: '/',
  redirectPathAfterSignOut: '/sign-in',
  adminRole: '20_example_admin',
};

config.signInAuthWrapper = creactUserIsNotAuthenticatedAuthWrapper(config.redirectPathAfterSignIn);
config.signUpAuthWrapper = creactUserIsNotAuthenticatedAuthWrapper(config.redirectPathAfterSignUp);

export const setConfig = (newConfig) => {
  Object.assign(config, newConfig);

  if (newConfig.redirectPathAfterSignIn) {
    updateSignInAuthWrapper();
  }

  if (newConfig.redirectPathAfterSignUp) {
    updateSignUpAuthWrapper();
  }
};

function updateSignInAuthWrapper() {
  config.signInAuthWrapper = creactUserIsNotAuthenticatedAuthWrapper(config.redirectPathAfterSignIn);
}

function updateSignUpAuthWrapper() {
  config.signUpAuthWrapper = creactUserIsNotAuthenticatedAuthWrapper(config.redirectPathAfterSignUp);
}

updateSignInAuthWrapper();
updateSignUpAuthWrapper();

export default config;
