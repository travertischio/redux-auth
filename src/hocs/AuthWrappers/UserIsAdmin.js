import { UserAuthWrapper } from 'redux-auth-wrapper';
import { routerActions } from 'react-router-redux';
// TODO: move it to redux-auth
import { selectUser } from 'redux-auth/lib/containers/AuthenticationProvider/selectors';

const UserIsAdmin = UserAuthWrapper({
  authSelector: selectUser,
  predicate: isAdmin,
  redirectAction: routerActions.replace,
  // TODO: move url of default homepage/dashboard to the config file
  failureRedirectPath: '/home',
  allowRedirectBack: false,
  wrapperDisplayName: 'UserIsAdmin',
});

function isAdmin(user) {
  return user && user.role === '20_example_admin';
}

export default UserIsAdmin;

