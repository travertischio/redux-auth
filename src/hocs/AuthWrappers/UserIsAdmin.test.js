import React from 'react';
import { createComponentWithRouter } from 'react-unit-testing-utils';
import config, { setConfig } from '~/config';
import UserIsAdmin from './UserIsAdmin';

const PageComponent = () => <div>Page only for admins</div>;
const PageOnlyForAdmins = UserIsAdmin(PageComponent);

describe('UserIsAdmin', () => {
  it('should be defined', () => {
    expect(UserIsAdmin).toBeDefined();
  });

  it('should redner PageComponent when user has admin role', () => {
    const initialState = {
      auth: {
        tokenData: {},
        userData: {
          role: config.adminRole,
        },
      },
    };

    const { wrapper } = createComponentWithRouter(PageOnlyForAdmins, initialState);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should not render PageComponent when user is not authenticated', () => {
    const initialState = {
      auth: {},
    };

    const { wrapper } = createComponentWithRouter(PageOnlyForAdmins, initialState);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should not render PageComponent when user has no admin role', () => {
    const initialState = {
      auth: {
        tokenData: {},
        userData: {
          role: '10_user',
        },
      },
    };

    const { wrapper } = createComponentWithRouter(PageOnlyForAdmins, initialState);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should not render PageComponent when user is not authenticated', () => {
    const initialState = {
      auth: {},
    };

    const { wrapper } = createComponentWithRouter(PageOnlyForAdmins, initialState);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render PageComponent when user has admin role and admin role is change in the config', () => {
    const newAdminRole = '100_superadmin';
    const initialState = {
      auth: {
        tokenData: {},
        userData: {
          role: newAdminRole,
        },
      },
    };

    setConfig({ adminRole: newAdminRole });

    const { wrapper } = createComponentWithRouter(PageOnlyForAdmins, initialState);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
