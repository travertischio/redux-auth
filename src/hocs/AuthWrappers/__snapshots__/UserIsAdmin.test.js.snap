// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserIsAdmin should display PageComponent when user has admin role 1`] = `
<div>
  Page only for admins
</div>
`;

exports[`UserIsAdmin when user has not admin role should display PageComponent when user has not admin role 1`] = `null`;
